### ISSUE TITLE
Updated web images do not get refreshed even with SDWebImageRefreshCached option

### ISSUE THREAD

**Issue by Srijeeth** [SDWebImage/SDWebImage#2813]
### New Issue Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / macos / tvos / watchos
 Platform Version        | e.g. 12.0 / 10.14.0 / 12.0 / 5.0
 SDWebImage Version      | e.g. 5.0.0 / 4.4.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 11 / Xcode 10
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

If Images are loaded in cache and updated or deleted in server, they do not get updated on the client side
I have used 
```objc
[cell.itemImageView sd_setImageWithURL:[NSURL URLWithString:path]                                      
                      placeholderImage:[UIImage imageNamed:@"nopreview.jpg"] 
                               options:SDWebImageRefreshCached
                             completed:nil];
```

**Comment by bpoplauschi** [comment]
@Srijeeth I have updated the title of this issue to a cleaner one and added the title you provided as description. Please fill in the issue template info that is required for us to understand more about this ... what version of our library did you use, ...

**Comment by dreampiggy** [comment]
This is designed behavior in the previous SDWebImage version. Your server can update the image file, and let the client to refresh. But there are no any sematic HTTP response, to `let client ignore the local cache, server think this should be removed`.

If you really need this syntax, maybe we need to find something standard way to take this. For example, server response a specify HTTP status code, or the HTTP response header contains this behavior.

**Comment by colinhumber** [comment]
@dreampiggy I'm running into a similar issue and it seems like a bug, but I'm not 100% sure.

I'm making a request to fetch an image using `SDWebImageManager` and providing the  `SDWebImageRefreshCached` flag. I'm also providing an `If-None-Match: "Etag"` header to the request. The expectation is that the cached image would be returned and the subsequent network request to fetch the refreshed image would return a 304. We see this behaviour in the browser. However, a 200 status code is always returned. 

After some investigation, it _looks_ like there's a bug in `SDWebImageDownloaderOperation` in the following condition:

```
if (self.options & SDWebImageDownloaderIgnoreCachedResponse) {
    // Grab the cached data for later check
    NSURLCache *URLCache = session.configuration.URLCache;
    if (!URLCache) {
        URLCache = [NSURLCache sharedURLCache];
    }
    NSCachedURLResponse *cachedResponse;
    // NSURLCache's `cachedResponseForRequest:` is not thread-safe, see https://developer.apple.com/documentation/foundation/nsurlcache#2317483
    @synchronized (URLCache) {
        cachedResponse = [URLCache cachedResponseForRequest:self.request];
    }
    if (cachedResponse) {
        self.cachedData = cachedResponse.data;
    }
}
```

Here, `self.options` contains the `SDWebImageDownloaderIgnoreCachedResponse` flag, so the assumption is it won't be considering the cached response.

**Comment by dreampiggy** [comment]
> However, a 200 status code is always returned.
I remember, URLSession will always **fake a 200** response code, if it actually hit the local cache using `NSURLRequestUseProtocolCachePolicy `. You can run a demo to check this behavior.

This does not effect the network transfer, it still respect HTTP Cache Policy, only the `NSHTTPURLResponse.statusCode` been faked.

**Comment by colinhumber** [comment]
Ahhh gotcha. Not the most obvious behaviour. I did see the comment in the operation where the 304 status code was checked and didn't quite understand it at first. That's good to know! Thanks for the quick reply!

**Comment by dreampiggy** [comment]
@colinhumber `-[NSURLCache cachedResponseForRequest:]` This does not always works as you expected. Sometimes it will actually, send a HTTP request to server (Server will response 304), this time we need that check.

But sometimes, URLSession can decide never to send HTTP request (when the last `Expire` does not hit), so it will return a local cached NSHTTPURLResponse (with faked 200 status code).

We have to handle both of these two :(

**Comment by SuiFengLiuNian** [comment]
这个方法始终不能响应相同URL的图片

**Comment by mrousavy** [comment]
> 这个方法始终不能响应相同URL的图片

yea

**Comment by mrousavy** [comment]
@dreampiggy I'm having the same problem right now, calling `sd_setImageWithURL` with `SDWebImageRefreshCached` but the image doesn't refresh even though the ETag is expired.

Should this work out of the box or do I need to manually tweak with passing the `If-None-Match` payload?

**Comment by mrousavy** [comment]
More context; Here's how I forceRefresh the image:

https://github.com/cuvent/react-native-fast-image/blob/3f247941bdf1627efa4fbe4dcd1506f50913d1d7/ios/FastImage/FFFastImageViewManager.m#L58-L68

⬇️ 

https://github.com/cuvent/react-native-fast-image/blob/3f247941bdf1627efa4fbe4dcd1506f50913d1d7/ios/FastImage/FFFastImageView.m#L119-L238
